applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/formfinding_strategy.hpp

```diff
@@ -0,0 +1,265 @@
+                    mWriteNewReferenceFile(WriteNewReferenceFile)
+                 {
+                    InitializeIterationIO();
+                 }
+
+            // constructor with Builder and Solver
+            FormfindingStrategy(
+                ModelPart& model_part,
+                typename TSchemeType::Pointer pScheme,
+                typename TLinearSolver::Pointer pNewLinearSolver,
+                typename TConvergenceCriteriaType::Pointer pNewConvergenceCriteria,
+                typename TBuilderAndSolverType::Pointer pNewBuilderAndSolver,
+                ModelPart& rFormFindingModelPart,
+                bool WriteNewReferenceFile,
+                std::string PrintingFormat,
+                Parameters ProjectionSetting,
+                int MaxIterations = 30,
+                bool CalculateReactions = false,
+                bool ReformDofSetAtEachStep = false,
+                bool MoveMeshFlag = false
+                )
+                : ResidualBasedNewtonRaphsonStrategy<TSparseSpace, TDenseSpace, TLinearSolver>(model_part, pScheme,
+                    pNewLinearSolver,pNewConvergenceCriteria,pNewBuilderAndSolver,MaxIterations,CalculateReactions,ReformDofSetAtEachStep,
+                    MoveMeshFlag),
+                    mProjectionSettings(ProjectionSetting),
+                    mrFormFindingModelPart(rFormFindingModelPart),
+                    mPrintingFormat(PrintingFormat),
+                    mWriteNewReferenceFile(WriteNewReferenceFile)
+                 {
+                    InitializeIterationIO();
+                 }
+
+            // Destructor
+            ~FormfindingStrategy() = default;
+
+        private:
+            bool SolveSolutionStep() override
+            {
+                BaseType::SolveSolutionStep();
+                if (mPrintingFormat=="all" || mPrintingFormat=="gid") mpIterationIO->FinalizeResults();
+                return true;
+            }
+
+            void UpdateDatabase(
+                    TSystemMatrixType& A,
+                    TSystemVectorType& Dx,
+                    TSystemVectorType& b,
+                    const bool MoveMesh
+                ) override
+                {
+                    BaseType::UpdateDatabase(A,Dx, b, MoveMesh);
+                    for(auto& r_node : mrFormFindingModelPart.Nodes()){
+                        // Updating reference
+                        const array_1d<double, 3>& disp = r_node.FastGetSolutionStepValue(DISPLACEMENT);
+                        array_1d<double, 3>& disp_non_historical = r_node.GetValue(DISPLACEMENT);
+
+                        disp_non_historical = disp_non_historical + disp;
+                        r_node.GetInitialPosition() += disp;
+
+                        r_node.FastGetSolutionStepValue(DISPLACEMENT) = ZeroVector(3);
+                    }
+                    ProjectVectorOnSurfaceUtility::Execute(mrFormFindingModelPart, mProjectionSettings);
+
+                    PrintResults();
```

Comment:
you cannot do this in EchoInfo?

Reply:
then I do not get the very last formfound result